### PR TITLE
#26 Fixes for ThreeD Secure completion

### PR DESCRIPTION
**Pull request by BLangendorf**

### CODE REVIEW

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java [hunk 1]
```diff
@@ -196,14 +202,27 @@ protected String getPaymentMethodId(final AdyenPaymentMethodsRecord paymentMetho
 
     @Override
     public PaymentTransactionInfoPlugin authorizePayment(final UUID kbAccountId, final UUID kbPaymentId, final UUID kbTransactionId, final UUID kbPaymentMethodId, final BigDecimal amount, final Currency currency, final Iterable<PluginProperty> properties, final CallContext context) throws PaymentPluginApiException {
+        final TransactionExecutor<PurchaseResult> authorizeTransactionExecutor;
+        if (hasPreviousAdyenRespondseRecord(kbPaymentId, kbTransactionId.toString(), context)) {
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: Maybe do the switch inside the executor (would simplify a bit)?

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java [hunk 2]
```diff
@@ -778,4 +829,26 @@ public boolean apply(final PaymentMethod paymentMethod) {
...
+        try {
+            final AdyenResponsesRecord previousAuthorizationResponse = dao.getSuccessfulAuthorizationResponse(kbPaymentId, context.getTenantId());
+            return previousAuthorizationResponse != null && previousAuthorizationResponse.getKbPaymentTransactionId().equals(kbPaymentTransactionId);
+        } catch (final SQLException e) {
+            logService.log(LogService.LOG_ERROR, "Failed to get previous AdyenResponsesRecord", e);
+            return false;
+        }
+    }
+
+    private Map<String, String[]> toRequestParameterMap(final Iterable<PluginProperty> pluginProperties) {
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: Can you use `PluginProperties.toStringMap` instead?

**Comment by pierre** [comment]
Could you also update the existing 3D-S test to verify we now effectively trigger the complete flow?

File: src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java (hunk 2 above)
Comment (BLangendorf) [inline src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: `PluginProperties.toStringMap` returns `Map<String, String>` but `adyenPort.authorize3DSecure()` expects `Map<String, String[]>`.

Comment (BLangendorf) [inline src/main/java/org/killbill/billing/plugin/adyen/api/AdyenPaymentPluginApi.java]: Maybe I can refactor the parameter and its usage to be `Map<String, String>` …

**Comment by BLangendorf** [comment]
@pierre The Test was wrong, did an independent purchase, I fixed that (300120d3957e609f59d912e77f0dece7b268e832) and now it really does a complete.
But I didn't find any hints in the resulting `PaymentTransactionInfoPlugin` (that it was a complete and not an authorize) _to verify we now effectively trigger the complete flow_

**Comment by BLangendorf** [comment]
@pierre Now with verification, that the `PspReference` of both transactions (initial auth, complete) are equal.

**Comment by pierre** [comment]
:tada:

**Pull request merged**